File: apps/guzo-api/src/main/java/et/guzo/service/FamilyService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.domain.entity.FamilyMember;
import et.guzo.domain.enums.FamilyRelation;
import et.guzo.repository.FamilyMemberRepository;
import et.guzo.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class FamilyService {

    private final FamilyMemberRepository familyMemberRepository;
    private final UserRepository userRepository;

    @Transactional(readOnly = true)
    public List<FamilyMember> listForOwner(String ownerUserId) {
        return familyMemberRepository.findByOwnerUserId(ownerUserId);
    }

    @Transactional
    public FamilyMember link(String ownerUserId, String memberUserId, FamilyRelation relation, String label) {
        if (ownerUserId.equals(memberUserId)) {
            throw ApiException.badRequest("Cannot link yourself as a family member");
        }
        if (!userRepository.existsById(ownerUserId) || !userRepository.existsById(memberUserId)) {
            throw ApiException.notFound("User not found");
        }
        if (familyMemberRepository.existsByOwnerUserIdAndMemberUserId(ownerUserId, memberUserId)) {
            throw ApiException.conflict("Family member already linked");
        }
        FamilyMember fm = new FamilyMember();
        fm.setId("fm_" + UUID.randomUUID().toString().replace("-", "").substring(0, 12));
        fm.setOwnerUserId(ownerUserId);
        fm.setMemberUserId(memberUserId);
        fm.setRelation(relation != null ? relation : FamilyRelation.OTHER);
        fm.setLabel(label);
        return familyMemberRepository.save(fm);
    }
}

File: apps/server/src/modules/family/family.service.ts
import { prisma } from '@delivery/database';
import { ApiError } from '../../utils/ApiError.js';

export class FamilyService {
  async listForOwner(ownerUserId: string) {
    const rows = await prisma.familyMember.findMany({
      where: { ownerUserId },
      orderBy: { createdAt: 'desc' },
      include: {
        member: {
          select: {
            firstName: true,
            lastName: true,
            phone: true,
            customer: { select: { customerCode: true } },
          },
        },
      },
    });

    return rows.map((row) => ({
      id: row.id,
      memberUserId: row.memberUserId,
      relation: row.relation,
      label: row.label,
      member: {
        firstName: row.member.firstName,
        lastName: row.member.lastName,
        phone: row.member.phone,
        guzoId: row.member.customer?.customerCode ?? row.memberUserId,
      },
    }));
  }

  async link(ownerUserId: string, memberUserId: string, relation: string, label?: string) {
    if (ownerUserId === memberUserId) {
      throw ApiError.badRequest('Cannot link yourself as a family member');
    }

    const [owner, member] = await Promise.all([
      prisma.user.findUnique({ where: { id: ownerUserId } }),
      prisma.user.findUnique({ where: { id: memberUserId } }),
    ]);
    if (!owner || !member) throw ApiError.notFound('User not found');

    const existing = await prisma.familyMember.findUnique({
      where: { ownerUserId_memberUserId: { ownerUserId, memberUserId } },
    });
    if (existing) throw ApiError.conflict('Family member already linked');

    const row = await prisma.familyMember.create({
      data: {
        ownerUserId,
        memberUserId,
        relation: (relation?.toUpperCase() || 'OTHER') as never,
        label: label?.trim() || null,
      },
      include: {
        member: {
          select: {
            firstName: true,
            lastName: true,
            phone: true,
            customer: { select: { customerCode: true } },
          },
        },
      },
    });

    return {
      id: row.id,
      memberUserId: row.memberUserId,
      relation: row.relation,
      label: row.label,
      member: {
        firstName: row.member.firstName,
        lastName: row.member.lastName,
        phone: row.member.phone,
        guzoId: row.member.customer?.customerCode ?? row.memberUserId,
      },
    };
  }
}

export const familyService = new FamilyService();
